File: frontend/lib/access/indexPlanAccess.mjs
// ---------------------------------------------------------------------------
// Index plan hierarchy — the ONE frontend entitlement authority.
//
// Basic -> Index Plus -> Index Premium. Premium satisfies every Plus check.
//
// AUTHENTICATION IS NOT ENTITLEMENT. Being signed in identifies a user; it
// grants nothing. An authenticated account with no paid plan has exactly the
// FEATURE access an anonymous visitor has. The two states are still kept apart
// wherever they are user-visible, because their upgrade paths differ (an
// anonymous visitor needs an account first), but no feature check may treat
// "signed in" as "entitled".
//
// The server half is `backend/domain/access/index_plan_access.py` and is an
// exact mirror. Neither side may invent its own reading of the same plans, and
// no surface may define a parallel notion of paid access — a second
// interpretation is how a paid feature quietly becomes free on one page.
// ---------------------------------------------------------------------------

export const INDEX_PLAN_PLUS = "plus";
export const INDEX_PLAN_PREMIUM = "premium";

export function normalizeIndexPlan(plan) {
  if (typeof plan !== "string") return null;
  const normalized = plan.trim().toLowerCase();
  return normalized === INDEX_PLAN_PLUS || normalized === INDEX_PLAN_PREMIUM
    ? normalized
    : null;
}

export function hasIndexPlusAccess(plan) {
  const normalized = normalizeIndexPlan(plan);
  return normalized === INDEX_PLAN_PLUS || normalized === INDEX_PLAN_PREMIUM;
}

export function hasIndexPremiumAccess(plan) {
  return normalizeIndexPlan(plan) === INDEX_PLAN_PREMIUM;
}

/**
 * Feature identity for Build a Market.
 *
 * Named for the CAPABILITY, not the plan. Commercial packaging is not final,
 * so the tier a feature maps to must be changeable in ONE place; hardcoding a
 * plan name at every call site is what makes repackaging a rewrite.
 */
export const FEATURE_MARKET_EXPLORER_CUSTOM_MARKETS = "market_explorer_custom_markets";
export const FEATURE_MARKET_EXPLORER_SINGLE_AXIS = "market_explorer_single_axis";
export const FEATURE_MARKET_EXPLORER_COMPOUND = "market_explorer_compound";
export const FEATURE_MARKET_EXPLORER_CUSTOM_RANKED = "market_explorer_custom_ranked";
export const FEATURE_MARKET_EXPLORER_POKEMON = "market_explorer_pokemon";
export const FEATURE_MARKET_EXPLORER_EXPLICIT_INSTRUMENTS = "market_explorer_explicit_instruments";
export const FEATURE_MARKET_EXPLORER_PREPARED_COMPARE = "market_explorer_prepared_compare";
export const FEATURE_MARKET_EXPLORER_ANALYTICAL_SCREENS = "market_explorer_analytical_screens";
export const FEATURE_MARKET_EXPLORER_ADVANCED_RANKING = "market_explorer_advanced_ranking";
export const FEATURE_CARD_CHASE_EFFICIENCY = "card_chase_efficiency";
export const FEATURE_PRODUCT_RIP = "product_rip";
export const FEATURE_DETAILED_OPENING_ECONOMICS = "detailed_opening_economics";
export const FEATURE_SET_PACK_ECONOMICS = "set_pack_economics";
export const FEATURE_ERA_PACK_ECONOMICS = "era_pack_economics";
export const FEATURE_MARKET_BREADTH = "market_breadth";
export const FEATURE_CARD_PULL_ODDS = "card_pull_odds";
export const FEATURE_ACQUISITION_MILESTONES = "acquisition_milestones";
export const FEATURE_PREPARED_MARKET_INTELLIGENCE = "prepared_market_intelligence";
export const FEATURE_CHASE_OPENING_ROUTE = "chase_opening_route";
export const FEATURE_CHASE_VS_BUY = "chase_vs_buy";
export const FEATURE_CHASE_RANKINGS = "chase_rankings";
// Product Chase Intelligence (Chase Access at Budget, O_budget) - a DISTINCT
// Premium construct from Card Chase Efficiency/Opening Route/vs Buy/Rankings
// above. "Which sealed product gives the most reach into a set's important
// value at my budget?", not "what's the best way to pursue this card?".
export const FEATURE_PRODUCT_CHASE_INTELLIGENCE = "product_chase_intelligence";

export const PLUS_FEATURES = Object.freeze(new Set([
  FEATURE_PRODUCT_RIP, FEATURE_DETAILED_OPENING_ECONOMICS,
  FEATURE_SET_PACK_ECONOMICS, FEATURE_ERA_PACK_ECONOMICS,
  FEATURE_MARKET_BREADTH, FEATURE_CARD_PULL_ODDS,
  FEATURE_ACQUISITION_MILESTONES, FEATURE_PREPARED_MARKET_INTELLIGENCE,
  FEATURE_MARKET_EXPLORER_PREPARED_COMPARE,
  FEATURE_MARKET_EXPLORER_ANALYTICAL_SCREENS,
  FEATURE_MARKET_EXPLORER_ADVANCED_RANKING,
]));
export const PREMIUM_FEATURES = Object.freeze(new Set([
  FEATURE_CARD_CHASE_EFFICIENCY, FEATURE_CHASE_OPENING_ROUTE,
  FEATURE_CHASE_VS_BUY, FEATURE_CHASE_RANKINGS,
  FEATURE_MARKET_EXPLORER_COMPOUND, FEATURE_MARKET_EXPLORER_CUSTOM_RANKED,
  FEATURE_MARKET_EXPLORER_POKEMON, FEATURE_MARKET_EXPLORER_EXPLICIT_INSTRUMENTS,
  FEATURE_MARKET_EXPLORER_CUSTOM_MARKETS, FEATURE_MARKET_EXPLORER_SINGLE_AXIS,
  FEATURE_PRODUCT_CHASE_INTELLIGENCE,
]));

export function hasIndexFeatureAccess(plan, feature) {
  if (PREMIUM_FEATURES.has(feature)) return hasIndexPremiumAccess(plan);
  if (PLUS_FEATURES.has(feature)) return hasIndexPlusAccess(plan);
  return false;
}

export function activeMarketFilterAxes(spec) {
  const axes = [];
  if (spec?.eraIds?.length || spec?.setIds?.length) axes.push("scope");
  if (spec?.segmentIds?.length) axes.push("segment");
  if (spec?.pokemonIds?.length) axes.push("pokemon");
  if (spec?.priceSegmentIds?.length) axes.push("priceSegment");
  if (spec?.releaseAgeCohortIds?.length) axes.push("releaseAge");
  return axes;
}

export function evaluateMarketQueryAccess(plan, spec) {
  const activeFilterAxes = activeMarketFilterAxes(spec);
  const ranked = spec?.mode === "chase";
  const pokemon = Boolean(spec?.pokemonIds?.length);
  const explicit = spec?.membershipMode === "explicit";
  // Every arbitrary Builder execution is creation, and creation is Premium.
  // Prepared-market browsing/comparison is evaluated by the Explorer
  // capability layer below and never reaches this query gate.
  const requiredPlan = INDEX_PLAN_PREMIUM;
  const capability = explicit
    ? FEATURE_MARKET_EXPLORER_EXPLICIT_INSTRUMENTS
    : pokemon
    ? FEATURE_MARKET_EXPLORER_POKEMON
    : ranked
      ? FEATURE_MARKET_EXPLORER_CUSTOM_RANKED
    : activeFilterAxes.length > 1
      ? FEATURE_MARKET_EXPLORER_COMPOUND
      : FEATURE_MARKET_EXPLORER_SINGLE_AXIS;
  const allowed = hasIndexPremiumAccess(plan);
  return { allowed, requiredPlan, capability, activeFilterAxes };
}

/** Plan display names. The product language, in one place. */
export const INDEX_PLAN_LABELS = Object.freeze({
  [INDEX_PLAN_PLUS]: "Index Plus",
  [INDEX_PLAN_PREMIUM]: "Index Premium",
});

/**
 * The Market Explorer access ladder for one user.
 *
 * THREE LEVELS, NOT TWO:
 *   basic   — Asset Market only (Raw, Sealed; Graded is unavailable to all).
 *   plus    — the prepared research layers: rarities, sealed families,
 *             Era & Sets browsing, benchmarks, prepared constituents.
 *   premium — everything above PLUS the custom-market builder.
 *
 * `accessMode` is "basic" for an anonymous visitor AND for an authenticated
 * account with no paid plan, because their FEATURE access is identical.
 * `isAuthenticated` is reported separately so the UI can offer the right next
 * step — sign in, or upgrade — without either being mistaken for entitlement.
 */
export function resolveMarketExplorerPlanAccess(user) {
  const indexPlan = normalizeIndexPlan(user?.index_plan);
  const canComparePreparedMarkets = hasIndexPlusAccess(indexPlan);
  const canBuildCustomMarkets = hasIndexPremiumAccess(indexPlan);
  return {
    accessMode: indexPlan || "basic",
    isAuthenticated: Boolean(user),
    indexPlan,
    // Plus AND Premium — Premium inherits the whole prepared layer.
    canBrowsePreparedMarkets: true,
    canComparePreparedMarkets,
    canUseAnalyticalScreens: canComparePreparedMarkets,
    canUseAdvancedMarketRanking: canComparePreparedMarkets,
    canUsePreparedMarketIntelligence: canComparePreparedMarkets,
    // Premium only. Browsing Era & Sets is a Plus capability; turning a scope
    // into a real custom market is not.
    canBuildCustomMarkets,
    canBuildSingleAxisMarket: canBuildCustomMarkets,
    canBuildCompoundMarket: hasIndexPremiumAccess(indexPlan),
    canUseCustomRankedComposition: hasIndexPremiumAccess(indexPlan),
  };
}

export function resolveRankingsPlanAccess(user) {
  const indexPlan = normalizeIndexPlan(user?.index_plan);
  return {
    canViewRankingsIntelligence: hasIndexPlusAccess(indexPlan),
    canViewCardChaseEfficiency: hasIndexFeatureAccess(indexPlan, FEATURE_CARD_CHASE_EFFICIENCY),
    accessMode: indexPlan || "basic",
  };
}

File: backend/domain/access/index_plan_access.py
"""Index plan hierarchy — the server-side half of the entitlement authority.

THE PLAN NAMES AND THE HIERARCHY ARE DEFINED ONCE PER RUNTIME. The browser half
lives in ``frontend/lib/access/indexPlanAccess.mjs`` and this is its exact
mirror: same two plan strings, same normalization, same "Premium satisfies
Plus" rule. A second interpretation of the same plans is how a paid feature
quietly becomes free on one surface, so neither side may invent its own.

FRONTEND GATES ARE PRESENTATION. This module exists so the API can refuse a
request that the UI merely declined to offer.
"""

from __future__ import annotations

from typing import Any, Mapping, Optional

INDEX_PLAN_PLUS = "plus"
INDEX_PLAN_PREMIUM = "premium"

#: The feature identity for Build a Market. Named for the CAPABILITY rather
#: than the plan, because commercial packaging is not final: if custom markets
#: later move to a differently-named tier, the seam is this constant's mapping
#: and not every call site.
FEATURE_MARKET_EXPLORER_CUSTOM_MARKETS = "market_explorer_custom_markets"
FEATURE_MARKET_EXPLORER_SINGLE_AXIS = "market_explorer_single_axis"
FEATURE_MARKET_EXPLORER_COMPOUND = "market_explorer_compound"
FEATURE_MARKET_EXPLORER_CUSTOM_RANKED = "market_explorer_custom_ranked"
FEATURE_MARKET_EXPLORER_POKEMON = "market_explorer_pokemon"
FEATURE_CARD_CHASE_EFFICIENCY = "card_chase_efficiency"
FEATURE_MARKET_BREADTH = "market_breadth"
FEATURE_PRODUCT_RIP = "product_rip"
FEATURE_PACK_ECONOMICS = "pack_economics"
FEATURE_ACQUISITION_MILESTONES = "acquisition_milestones"
FEATURE_SET_RIP_ANALYTICS = "set_rip_analytics"

_PLUS_FEATURES = frozenset({
    FEATURE_MARKET_EXPLORER_CUSTOM_MARKETS,
    FEATURE_MARKET_EXPLORER_SINGLE_AXIS,
    FEATURE_MARKET_BREADTH,
    FEATURE_PRODUCT_RIP,
    FEATURE_PACK_ECONOMICS,
    FEATURE_ACQUISITION_MILESTONES,
    FEATURE_SET_RIP_ANALYTICS,
})
_PREMIUM_FEATURES = frozenset({
    FEATURE_MARKET_EXPLORER_COMPOUND,
    FEATURE_MARKET_EXPLORER_CUSTOM_RANKED,
    FEATURE_MARKET_EXPLORER_POKEMON,
    FEATURE_CARD_CHASE_EFFICIENCY,
})


def normalize_index_plan(plan: Any) -> Optional[str]:
    if not isinstance(plan, str):
        return None
    normalized = plan.strip().lower()
    return normalized if normalized in (INDEX_PLAN_PLUS, INDEX_PLAN_PREMIUM) else None


def has_index_plus_access(plan: Any) -> bool:
    """True for Index Plus AND Index Premium — Premium inherits Plus."""
    return normalize_index_plan(plan) in (INDEX_PLAN_PLUS, INDEX_PLAN_PREMIUM)


def has_index_premium_access(plan: Any) -> bool:
    return normalize_index_plan(plan) == INDEX_PLAN_PREMIUM


def has_index_feature_access(plan: Any, feature: str) -> bool:
    """Canonical capability-to-plan mapping; unknown capabilities fail closed."""
    if feature in _PREMIUM_FEATURES:
        return has_index_premium_access(plan)
    if feature in _PLUS_FEATURES:
        return has_index_plus_access(plan)
    return False


def evaluate_market_query_access(plan: Any, spec: Mapping[str, Any]) -> dict[str, Any]:
    """Evaluate one normalized query definition against the plan hierarchy."""
    from backend.domain.pokemon.market_explorer_query import MODE_CHASE, active_filter_axes

    axes = active_filter_axes(spec)
    ranked = spec.get("mode") == MODE_CHASE
    if spec.get("pokemonIds"):
        required_plan, capability = INDEX_PLAN_PREMIUM, FEATURE_MARKET_EXPLORER_POKEMON
        reason = "Pokemon market research requires Index Premium"
    elif ranked:
        required_plan, capability = INDEX_PLAN_PREMIUM, FEATURE_MARKET_EXPLORER_CUSTOM_RANKED
        reason = "custom ranked composition requires Index Premium"
    elif len(axes) > 1:
        required_plan, capability = INDEX_PLAN_PREMIUM, FEATURE_MARKET_EXPLORER_COMPOUND
        reason = "compound market requires Index Premium"
    else:
        required_plan, capability = INDEX_PLAN_PLUS, FEATURE_MARKET_EXPLORER_SINGLE_AXIS
        reason = "custom market exploration requires Index Plus"
    allowed = (has_index_premium_access(plan) if required_plan == INDEX_PLAN_PREMIUM
               else has_index_plus_access(plan))
    return {
        "allowed": allowed,
        "requiredPlan": required_plan,
        "capability": capability,
        "reason": None if allowed else reason,
        "activeFilterAxes": list(axes),
    }


def _pick(source: Mapping[str, Any], allowed: frozenset[str]) -> dict[str, Any]:
    """Copy only fields classified for a response contract.

    This is intentionally not a redactor. New upstream fields are absent until
    somebody adds them to a tier allowlist.
    """
    return {key: source[key] for key in allowed if key in source}


_BASE_TARGET_FIELDS = frozenset({
    "id", "set_id", "target_id", "target_type", "name", "slug", "canonical_key",
    "pokemon_api_set_id", "era", "era_id", "hero_image_url", "logo_image_url",
    "symbol_image_url", "checklist_set_value", "checklist_set_value_as_of",
    "checklist_set_value_priced_card_count", "checklist_set_value_total_card_count",
    "checklistSetValue", "checklistSetValueAsOf", "checklistSetValuePricedCardCount",
    "checklistSetValueTotalCardCount", "current_checklist_set_value",
    "current_checklist_set_value_date", "currentChecklistSetValue",
    "currentChecklistSetValueDate", "publicAnalyticsStatus",
})
_PLUS_TARGET_FIELDS = _BASE_TARGET_FIELDS | frozenset({
    "calculation_run_id", "run_at", "pack_cost", "pack_score", "relative_pack_score",
    "pack_rank", "pack_tier", "profit_score", "relative_profit_score", "profit_rank",
    "profit_tier", "safety_score", "relative_safety_score", "safety_rank", "safety_tier",
    "stability_score", "relative_stability_score", "stability_rank", "stability_tier",
    "mean_value", "median_value", "prob_profit", "prob_big_hit", "roi_percent",
    "expected_loss_when_losing", "mean_value_to_cost_ratio", "mean_value_to_cost_rank",
    "mean_value_to_cost_tier", "p95_value_to_cost_ratio", "p95_value_to_cost_rank",
    "p95_value_to_cost_tier", "p99_value_to_cost_ratio", "p99_value_to_cost_rank",
    "p99_value_to_cost_tier", "rip", "ripCore", "financialRipV4", "overallRipV10",
    "publicRipContractV10", "setRipV1", "openingExperience", "rankingsChase",
    "collector_appeal_score", "collector_appeal_rank", "opening_desirability_score",
    "opening_desirability_rank", "opening_desirability_summary",
})
_RANKINGS_META_FIELDS = frozenset({"source", "updatedAt", "warnings", "snapshot", "limit"})
_PUBLIC_SET_RIP_FIELDS = frozenset({
    "score", "rank", "tier", "cohortSize", "rankable", "methodologyVersion",
    "participatingFamilyCount", "participatingFamilies", "skuEvidenceCount",
    "familyScores", "displayFamilyScores",
})


def _project_public_set_leaderboard_target(target: Mapping[str, Any]) -> dict[str, Any]:
    projected = _pick(target, _BASE_TARGET_FIELDS)
    set_rip = target.get("setRipV1")
    if isinstance(set_rip, Mapping):
        projected["setRipV1"] = _pick(set_rip, _PUBLIC_SET_RIP_FIELDS)
    return projected


def project_rankings_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    """Tier-safe `/explore/rip-statistics/targets` response."""
    plus = has_index_feature_access(plan, FEATURE_SET_RIP_ANALYTICS)
    target_fields = _PLUS_TARGET_FIELDS if plus else _BASE_TARGET_FIELDS
    targets = [
        _pick(target, target_fields) if plus else _project_public_set_leaderboard_target(target)
        for target in payload.get("targets", []) if isinstance(target, Mapping)
    ]
    result: dict[str, Any] = {
        "targets": targets,
        "meta": _pick(payload.get("meta") or {}, _RANKINGS_META_FIELDS),
        "access": {"rankingsIntelligence": plus, "requiredPlan": "plus"},
    }
    default_target = payload.get("default_target") or payload.get("defaultTarget")
    if isinstance(default_target, Mapping):
        result["default_target"] = (
            _pick(default_target, target_fields) if plus
            else _project_public_set_leaderboard_target(default_target)
        )
    if plus:
        for key in ("setRip", "productFamilyRankings"):
            if key in payload:
                result[key] = payload[key]
    return result


_PUBLIC_ERA_FIELDS = frozenset({
    "eraName", "eraId", "rank", "score", "tier", "cohortSize",
    "modeledSetCount", "strongestSet", "constituentSets",
})
_PUBLIC_ERA_SET_FIELDS = frozenset({"setId", "setName", "score"})


def project_public_era_rankings_response(payload: Mapping[str, Any]) -> dict[str, Any]:
    """Expose only the prepared fields rendered by the public Era leaderboard."""
    source = payload.get("eraSetStrengthV1")
    if not isinstance(source, Mapping):
        return {"cohortSize": 0, "eras": []}
    eras = []
    for row in source.get("eras", []):
        if not isinstance(row, Mapping):
            continue
        projected = _pick(row, _PUBLIC_ERA_FIELDS)
        strongest = row.get("strongestSet")
        projected["strongestSet"] = _pick(strongest, _PUBLIC_ERA_SET_FIELDS) if isinstance(strongest, Mapping) else None
        projected["constituentSets"] = [
            _pick(item, _PUBLIC_ERA_SET_FIELDS)
            for item in row.get("constituentSets", []) if isinstance(item, Mapping)
        ]
        eras.append(projected)
    return {"cohortSize": source.get("cohortSize"), "eras": eras}


_BASE_PRODUCT_RANKING_FIELDS = frozenset({
    "sealedProductId", "setId", "productName", "setName", "productFamily",
    "productFamilyLabel", "productImageUrl", "setCanonicalKey", "unitPrice", "marketPrice",
})
_PLUS_PRODUCT_RANKING_FIELDS = _BASE_PRODUCT_RANKING_FIELDS | frozenset({
    "budgetRank", "budgetCohortSize", "budgetTier", "budgetModelTier", "publicTier",
    "quantity", "actualCommittedCapital", "unusedCapital", "overallRipScore",
    "financialRipScore", "overallRipAbsoluteScore", "overallRipRelativeScore",
    "overallRipLeaderScore", "financialRipAbsoluteScore", "financialRipRelativeScore",
    "financialRipLeaderScore", "collectorAppealScore", "expectedValue",
    "chanceToRecoverCost", "familyRank", "familySize", "familyTier",
})


def project_product_rankings_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    plus = has_index_feature_access(plan, FEATURE_PRODUCT_RIP)
    fields = _PLUS_PRODUCT_RANKING_FIELDS if plus else _BASE_PRODUCT_RANKING_FIELDS
    return {
        "available": bool(payload.get("available")),
        "reason": payload.get("reason"),
        "selectedBudget": payload.get("selectedBudget"),
        "availableBudgets": payload.get("availableBudgets") or [],
        "cohortSize": payload.get("cohortSize", 0),
        "rows": [_pick(row, fields) for row in payload.get("rows", []) if isinstance(row, Mapping)],
        **({"authority": payload.get("authority") or {}} if plus else {}),
    }


def project_product_family_rankings_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    """Tier-safe nested Product Family Rankings publication."""
    plus = has_index_feature_access(plan, FEATURE_PRODUCT_RIP)
    fields = _PLUS_PRODUCT_RANKING_FIELDS if plus else _BASE_PRODUCT_RANKING_FIELDS
    source_families = payload.get("families") or {}
    families: dict[str, Any] = {}
    if isinstance(source_families, Mapping):
        for family_id, block in source_families.items():
            if not isinstance(block, Mapping):
                continue
            products = block.get("products") or block.get("rows") or []
            projected = _pick(block, frozenset({"label", "count", "productFamily", "productFamilyLabel"}))
            projected["products"] = [
                _pick(row, fields) for row in products if isinstance(row, Mapping)
            ]
            families[str(family_id)] = projected
    result = {"families": families}
    if plus:
        result.update(_pick(payload, frozenset({"authority", "authorityTargetCount"})))
    return result


_OPENING_SCOPE_FIELDS = frozenset({
    "averageCostPerPack", "averageModelBreakEvenPerPack", "averageEntertainmentCostPerPack",
    "modeledReturnOnSpend", "entertainmentCostShare", "meanOutcomeRetention",
    "chanceToRecoverCost", "typicalOpeningPerPack", "typicalRetention",
    "valuePerPackPercentiles", "normalizedReturnPercentiles", "setCount",
    "productSkuCount", "productFamilyCount",
})
_BASE_OPENING_BREAKDOWN_FIELDS = frozenset({
    "eraId", "eraName", "setId", "setName", "setCanonicalKey",
    "setCount", "productSkuCount", "productFamilyCount", "averageCostPerPack",
})


def project_opening_economics_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    """Keep the global educational contract public; tier detailed breakdowns."""
    plus = has_index_feature_access(plan, FEATURE_PACK_ECONOMICS)
    result = _pick(payload, frozenset({
        "status", "reason", "contractVersion", "basis", "methodology", "marketDate", "population",
    }))
    if isinstance(payload.get("global"), Mapping):
        result["global"] = _pick(payload["global"], _OPENING_SCOPE_FIELDS)
    else:
        result["global"] = None
    for key in ("eras", "sets"):
        result[key] = []
        for row in payload.get(key, []):
            if not isinstance(row, Mapping):
                continue
            projected = _pick(
                row,
                _OPENING_SCOPE_FIELDS | _BASE_OPENING_BREAKDOWN_FIELDS
                if plus else _BASE_OPENING_BREAKDOWN_FIELDS,
            )
            if plus and key == "sets":
                projected["familyEconomics"] = [
                    _pick(family, _OPENING_SCOPE_FIELDS | frozenset({"family"}))
                    for family in row.get("familyEconomics", []) if isinstance(family, Mapping)
                ]
            result[key].append(projected)
    result["familyBenchmarks"] = [
        _pick(row, _OPENING_SCOPE_FIELDS | frozenset({"family"}))
        for row in payload.get("familyBenchmarks", []) if plus and isinstance(row, Mapping)
    ]
    return result


_BASE_MARKET_TOP_LEVEL = frozenset({
    "set", "window", "window_key", "days", "latestMarketDate", "latest_market_date",
    "availableScopes", "available_scopes", "setValueHistoriesByScope",
    "set_value_histories_by_scope", "performanceVsCostHistory",
    "performance_vs_cost_history", "topChaseCards", "top_chase_cards",
    "topChaseCardHistories", "top_chase_card_histories", "marketMovers",
    "market_movers", "marketMoversByWindow", "market_movers_by_window", "meta",
})


def project_set_market_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    result = _pick(payload, _BASE_MARKET_TOP_LEVEL)
    for key in ("cardsMarket", "cards_market"):
        block = payload.get(key)
        if not isinstance(block, Mapping):
            continue
        safe = _pick(block, frozenset({
            "available", "status", "marketIndex", "market_index", "setValue",
            "set_value", "history", "asOf", "as_of", "source",
        }))
        if has_index_feature_access(plan, FEATURE_MARKET_BREADTH):
            for breadth_key in ("marketBreadth", "market_breadth"):
                if breadth_key in block:
                    safe[breadth_key] = block[breadth_key]
        result[key] = safe
    for contract_key in ("publicRipContractV10", "public_rip_contract_v10"):
        contract = payload.get(contract_key)
        if not isinstance(contract, Mapping):
            continue
        collector = contract.get("collectorAppeal") or contract.get("collector_appeal")
        if not isinstance(collector, Mapping):
            continue
        projected_subjects = []
        for subject in collector.get("topSubjects") or collector.get("top_subjects") or []:
            if not isinstance(subject, Mapping):
                continue
            projected_subject = _pick(subject, frozenset({
                "subjectName", "subject_name", "subjectId", "subject_id", "imageUrl", "image_url",
            }))
            for path_key in ("elitePath", "elite_path", "accessiblePath", "accessible_path"):
                path = subject.get(path_key)
                if not isinstance(path, Mapping):
                    continue
                allowed = {
                    "canonicalCardId", "canonical_card_id", "cardName", "card_name",
                    "imageUrl", "image_url", "modeledProbability", "modeled_probability",
                    "impliedOdds", "implied_odds",
                }
                if has_index_feature_access(plan, FEATURE_ACQUISITION_MILESTONES):
                    allowed.update({
                        "packsFor50PercentChance", "packs_for_50_percent_chance",
                        "packsFor90PercentChance", "packs_for_90_percent_chance",
                        "approximatePackSpend50", "approximate_pack_spend_50",
                        "approximatePackSpend90", "approximate_pack_spend_90",
                    })
                projected_subject[path_key] = _pick(path, frozenset(allowed))
            projected_subjects.append(projected_subject)
        result[contract_key] = {
            "collectorAppeal": {"topSubjects": projected_subjects}
        }
    return result


_BASE_SET_PAGE_FIELDS = frozenset({"target", "set", "meta"})
_PLUS_SET_PAGE_FIELDS = _BASE_SET_PAGE_FIELDS | frozenset({
    "summary", "interpretation", "rankings", "ripDecision", "rip_statistics",
    "percentiles", "distribution_bins", "threshold_bins", "history_trend",
    "top_hits", "openingDesirability", "pull_rate_assumptions",
    "cardDesirabilityValidation", "card_desirability_validation",
    "cardAppealMarketPriceCorrelation", "card_appeal_market_price_correlation",
    "rip", "ripCore", "financialRipV4", "overallRipV10", "publicRipContractV10",
    "openingExperience", "publicAnalyticsCohort", "publicAnalyticsStatus",
})


def project_set_page_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    fields = _PLUS_SET_PAGE_FIELDS if has_index_feature_access(plan, FEATURE_SET_RIP_ANALYTICS) else _BASE_SET_PAGE_FIELDS
    return _pick(payload, fields)


def project_card_detail_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    result = _pick(payload, frozenset({
        "set", "card", "availableVariants", "selectedVariantId", "variantSelection",
        "market", "meta",
    }))
    if has_index_feature_access(plan, FEATURE_PRODUCT_RIP) and "intelligence" in payload:
        result["intelligence"] = payload["intelligence"]
    if has_index_feature_access(plan, FEATURE_PRODUCT_RIP) and "treatmentMarketPrestige" in payload:
        result["treatmentMarketPrestige"] = payload["treatmentMarketPrestige"]
    if has_index_feature_access(plan, FEATURE_CARD_CHASE_EFFICIENCY) and "chase" in payload:
        result["chase"] = payload["chase"]
    return result


def project_sealed_product_detail_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    result = _pick(payload, frozenset({"set", "product", "market", "meta"}))
    if has_index_feature_access(plan, FEATURE_PRODUCT_RIP):
        result.update(_pick(payload, frozenset({"rip", "comparisons"})))
    return result


def project_sealed_market_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    result = _pick(payload, frozenset({
        "set", "marketDate", "defaultProductId", "setPageConsumerMarket",
        "setPageConsumerTopProducts", "meta",
    }))
    if has_index_feature_access(plan, FEATURE_MARKET_BREADTH):
        result.update(_pick(payload, frozenset({"products", "setMarket"})))
    return result


def project_insights_critical_response(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    if has_index_feature_access(plan, FEATURE_SET_RIP_ANALYTICS):
        return _pick(payload, frozenset({
            "ripDecision", "set", "summary", "recommendation", "ripScore", "rip",
            "ripCore", "financialRipV4", "overallRipV10", "publicRipContractV10",
            "openingExperience", "publicAnalyticsCohort", "publicAnalyticsStatus",
            "interpretation", "meta",
        }))
    return _pick(payload, frozenset({
        "set", "recommendation", "ripScore", "publicAnalyticsStatus", "meta",
    }))


def filter_set_market_signal_access(payload: Mapping[str, Any], plan: Any) -> dict[str, Any]:
    """Compatibility name for the allowlist-based market projector."""
    return project_set_market_response(payload, plan)


def resolve_market_explorer_plan_access(user: Mapping[str, Any] | None) -> dict[str, Any]:
    """The Market Explorer access ladder for one user.

    ``accessMode`` is "basic" for both an anonymous visitor and an authenticated
    user with no paid plan: they have the same FEATURE access. Authentication
    and entitlement stay separate states everywhere they are user-visible (the
    upgrade path differs), but nothing about what the API will serve turns on
    being signed in alone.
    """
    plan = normalize_index_plan((user or {}).get("index_plan"))
    return {
        "accessMode": plan or "basic",
        "canUsePreparedMarketIntelligence": has_index_plus_access(plan),
        "canBuildCustomMarkets": has_index_plus_access(plan),
        "canBuildSingleAxisMarket": has_index_plus_access(plan),
        "canBuildCompoundMarket": has_index_premium_access(plan),
        "canUseCustomRankedComposition": has_index_premium_access(plan),
    }
